File: app/api/v1/endpoints/social_media.py
from fastapi import APIRouter, HTTPException, Depends
from typing import Dict, Any, Optional
from datetime import datetime
import logging

from app.services.social_media.instagram_service import instagram_service
from app.core.deps import get_current_user
from app.models.user import User

logger = logging.getLogger(__name__)
router = APIRouter()

@router.get("/instagram/metrics")
async def get_instagram_metrics(
    current_user: User = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Get Instagram metrics for OKR tracking.
    Requires Instagram API access token to be configured.
    """
    try:
        metrics = await instagram_service.get_comprehensive_metrics()
        
        if metrics["status"] == "error":
            raise HTTPException(
                status_code=503,
                detail=f"Instagram API error: {metrics.get('error', 'Unknown error')}"
            )
        
        return {
            "success": True,
            "data": metrics,
            "timestamp": datetime.utcnow().isoformat()
        }
        
    except Exception as e:
        logger.error(f"Error fetching Instagram metrics: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Failed to fetch Instagram metrics"
        )

@router.get("/instagram/followers")
async def get_instagram_followers(
    current_user: User = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Get current Instagram follower count.
    """
    try:
        follower_count = await instagram_service.get_follower_count()
        
        if follower_count is None:
            raise HTTPException(
                status_code=503,
                detail="Unable to fetch Instagram follower count. Check API configuration."
            )
        
        return {
            "success": True,
            "data": {
                "platform": "instagram",
                "followers": follower_count,
                "timestamp": datetime.utcnow().isoformat()
            }
        }
        
    except Exception as e:
        logger.error(f"Error fetching Instagram followers: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Failed to fetch Instagram follower count"
        )

@router.get("/instagram/account")
async def get_instagram_account_info(
    current_user: User = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Get Instagram account information.
    """
    try:
        account_info = await instagram_service.get_account_info()
        
        if account_info is None:
            raise HTTPException(
                status_code=503,
                detail="Unable to fetch Instagram account info. Check API configuration."
            )
        
        return {
            "success": True,
            "data": account_info
        }
        
    except Exception as e:
        logger.error(f"Error fetching Instagram account info: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Failed to fetch Instagram account information"
        )

@router.get("/instagram/media")
async def get_instagram_recent_media(
    limit: int = 10,
    current_user: User = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Get recent Instagram media posts with engagement metrics.
    """
    try:
        media = await instagram_service.get_recent_media(limit=limit)
        
        if media is None:
            raise HTTPException(
                status_code=503,
                detail="Unable to fetch Instagram media. Check API configuration."
            )
        
        return {
            "success": True,
            "data": media
        }
        
    except Exception as e:
        logger.error(f"Error fetching Instagram media: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail="Failed to fetch Instagram media"
        )

@router.get("/social-media/status")
async def get_social_media_status(
    current_user: User = Depends(get_current_user)
) -> Dict[str, Any]:
    """
    Get status of all social media API connections.
    """
    status = {
        "instagram": {
            "configured": bool(instagram_service.access_token),
            "business_account_id": bool(instagram_service.business_account_id),
            "status": "configured" if instagram_service.access_token else "not_configured"
        },
        "overall_status": "partial" if instagram_service.access_token else "not_configured"
    }
    
    return {
        "success": True,
        "data": status,
        "timestamp": datetime.utcnow().isoformat()
    } 

File: frontend/src/services/social-media.ts
import { apiClient } from '@/lib/api-client';

export interface InstagramMetrics {
  platform: string;
  timestamp: string;
  followers: number | null;
  media_count: number | null;
  recent_engagement: {
    total_likes: number;
    total_comments: number;
    avg_likes_per_post: number;
    avg_comments_per_post: number;
  } | null;
  account_info: {
    username: string;
    biography: string;
    profile_picture: string;
  } | null;
  status: string;
  error: string | null;
}

export interface InstagramAccountInfo {
  id: string;
  username: string;
  followers_count: number;
  media_count: number;
  biography: string;
  profile_picture_url: string;
}

export interface InstagramMedia {
  id: string;
  caption: string;
  media_type: string;
  media_url: string;
  thumbnail_url?: string;
  permalink: string;
  timestamp: string;
  like_count: number;
  comments_count: number;
}

export interface SocialMediaStatus {
  instagram: {
    configured: boolean;
    business_account_id: boolean;
    status: string;
  };
  overall_status: string;
}

export class SocialMediaService {
  /**
   * Get comprehensive Instagram metrics for OKR tracking
   */
  static async getInstagramMetrics(): Promise<InstagramMetrics> {
    try {
      const response = await apiClient.get('/social-media/instagram/metrics');
      return response.data.data;
    } catch (error) {
      console.error('Error fetching Instagram metrics:', error);
      throw new Error('Failed to fetch Instagram metrics');
    }
  }

  /**
   * Get current Instagram follower count
   */
  static async getInstagramFollowers(): Promise<number> {
    try {
      const response = await apiClient.get('/social-media/instagram/followers');
      return response.data.data.followers;
    } catch (error) {
      console.error('Error fetching Instagram followers:', error);
      throw new Error('Failed to fetch Instagram follower count');
    }
  }

  /**
   * Get Instagram account information
   */
  static async getInstagramAccountInfo(): Promise<InstagramAccountInfo> {
    try {
      const response = await apiClient.get('/social-media/instagram/account');
      return response.data.data;
    } catch (error) {
      console.error('Error fetching Instagram account info:', error);
      throw new Error('Failed to fetch Instagram account information');
    }
  }

  /**
   * Get recent Instagram media posts
   */
  static async getInstagramRecentMedia(limit: number = 10): Promise<InstagramMedia[]> {
    try {
      const response = await apiClient.get(`/social-media/instagram/media?limit=${limit}`);
      return response.data.data.data || [];
    } catch (error) {
      console.error('Error fetching Instagram media:', error);
      throw new Error('Failed to fetch Instagram media');
    }
  }

  /**
   * Get status of all social media API connections
   */
  static async getSocialMediaStatus(): Promise<SocialMediaStatus> {
    try {
      const response = await apiClient.get('/social-media/status');
      return response.data.data;
    } catch (error) {
      console.error('Error fetching social media status:', error);
      throw new Error('Failed to fetch social media status');
    }
  }

  /**
   * Test Instagram API connection
   */
  static async testInstagramConnection(): Promise<boolean> {
    try {
      await this.getInstagramFollowers();
      return true;
    } catch (error) {
      return false;
    }
  }
} 